### ISSUE TITLE
Per-Layer Embedding (Google) ~ Engram (Deep Seek/Meituan) ~ OverEncoding (Meituan) ~ Sparse update rule/Optimization state is massive.

### ISSUE THREAD

**Issue by CoffeeVampir3** [huggingface/peft#3371]
### Feature request

Engrams/Per-Layer embedding/Overencoding -- all reminiscent of the same idea. These are very likely to become an integral part of any new transformer, alongside MoE and attention layers. Currently, as far as I can tell, PEFT doesn't have much of a strategy for these. At low batch sizes these updates are typically sparse, if we have e.g. 1 million rows and a batch size of 40, with 8000 context length, we actiavate maximally 32k rows per batch. 

For just gemma 2E2 ~ it looks like around 30GB of adam optimizer state just for the embedding tables. That's non-transient, so even sparse updates would eventually accumulate the same memory. I'm not sure what the typical case is, but PEFT may want to special case engram-like embedding tables to use a row-wise optimizer like RMSprop being per-row in this case would be something like ~230MB or so (with a worse outcome, but a much more tractible default I think) ~ Combined with sparse updates to avoid materializing the entire embedding table intermediate activation, I think this would make training engram-like models much more realistic.

https://huggingface.co/google/gemma-4-E2B
https://huggingface.co/meituan-longcat/LongCat-Flash-Lite
https://arxiv.org/abs/2601.07372
https://arxiv.org/abs/2601.21204

### Your contribution

I have my own bias-correct RMSProp sparse optimizer for these specific engrams here <https://github.com/CoffeeVampir3/Architecture-Zarelga-Superposition/blob/main/engram/GramReaperSparse.py>

And a sparse-engram setup here: <https://github.com/CoffeeVampir3/Architecture-Zarelga-Superposition/blob/main/engram/EngramEmbeddingSparse.py>

Using a bespoke optimizer is probably too much, but any row-wise optimizer is a much more realistic choice here. There's also sage (much more expensive than RMSprop) which cuts the memory in half and is otherwise largely similar to lion. <https://arxiv.org/abs/2604.07663>

My own measurements show that adam is objectively the better optimizer VS sage, and both of these are better than RMSprop, so there's certainly a large memory | quality tradeoff.

On a synthetic zipf-distributed head/tail recall accuracy measure that trained an engram module on a synthetic recall task:

<img width="410" height="168" alt="Image" src="https://github.com/user-attachments/assets/703a408e-6ce7-4549-9305-e67dd849db9e" />

Lion and sage are both sensitive clearly to LR so these don't rule out those optimizers, but they are 1/2 memory-ish of adam as opposed to massively less. Also appears to be somewhat inconsistent seed-to-seed VS the others.

<img width="353" height="138" alt="Image" src="https://github.com/user-attachments/assets/a707837e-942b-4c57-a6be-d2594af5c3fd" />

TLDR: Current treatment of engram/PLE is not very parameter-efficient in terms of memory.

**Comment by githubnemo** [comment]
Hey, thanks for the heads-up.

Indeed, the PLE embedding layer is quite memory intensive when trained directly (e.g., `modules_to_save=["embed_tokens_per_layer"]`). PEFT does give you three options, though:

1. target `embed_tokens_per_layer` with a PEFT method (e.g., LoRA) - the optimizer state will be a lot smaller in this case but the update will be global, not token-local (which is arguably the main benefit of PLE)
2. target the layer-local PLE linear projection - this is better than (1) in the sense that it is now layer-local but it will still be token-global. It might very well be that this is comparable to fully tuning the `embed_tokens_per_layer` matrix
3. determine a selection of tokens (e.g., top-k scores via tf-idf on the training dataset) and use `trainable_tokens_indices={"embed_tokens_per_layer": <token_ids>}` to selectively train only those tokens.

I think it would be very interesting to see how these approaches fare against each other in, say, the PEFT MetaMathQA benchmark before implementing a fourth option, so that we have a proper baseline. Would you be interested in doing this?

Regarding the proposed fourth option: I like the approach of having a specific optimizer for the PLE embedding layer but it is not clear to me whether something like using SGD for that layer is sufficient. Did you test this already?

**Edit:** I quickly tested gemma4-e2b + LoRA on MetaMathQA, one run with default setting and one run with additionally setting `modules_to_save=["embed_tokens_per_layer"]` and the first one takes ~36GB VRAM and the second one ~47GB VRAM before OOMing so the general hypothesis that targeting PLE is memory intensive is confirmed.
